docs/proposals/controller/README.md

```diff
@@ -0,0 +1,58 @@
+# Cluster Controller
+
+The Cluster Controller is an intermediary between the client and scheduler, its main task is to compare two states: the one requested by the client and the existing one reported by the scheduler and saved in the database. Based on the comparison results, the controller sends requests to the scheduler API in order to transform the current state into the requested one.
+
+### Services provided:
+
+HTTP API (OpenAPI 3.0 compliant) to connect with CLI
+gRPC client and API for the scheduler
+
+### Scope
+
+The Cluster Controller is responsible for:
+
+- Receiving requests from the CLI (Controller External API)
+
+- Receive updates on the instances status from scheduler (Controller Internal API)
+- Store current cluster state in the database
+- Compare current state with the requested one when an update request is received from a client or when the scheduler reports a change in the cluster state.
+- Make requests to scheduler based on the comparison results
```

Comment:
I don't think this is correct. The controller should not make any decision, either the end user (the external API client) or the scheduler make decisions; not the controller.

The controller receives requests from a client and reaches the scheduler based on those requests. It then *tracks* the instances and nodes statuses and reports that to end users. Those then decide how to proceed, depending on that piece of information.

Reply:
yes, makes more sense

Reply:
Not for me. I get it that we shouldn't try to decide how to fix problems when receiving updates from the scheduler, but what about client requests? Scheduler doesn't have access to the database, they won't be able to compare states and determine if a demanded workload already exists. Shouldn't we at least make the decision of what needs to be given to scheduler, and what is deployed already and should remain unchanged?

If not, does it mean that the controller serves only to pass workloads to scheduler, store data about states/workloads and give this data to client?

Reply:
> Not for me. I get it that we shouldn't try to decide how to fix problems when receiving updates from the scheduler, but what about client requests? Scheduler doesn't have access to the database, they won't be able to compare states and determine if a demanded workload already exists. Shouldn't we at least make the decision of what needs to be given to scheduler, and what is deployed already and should remain unchanged?
> 

The client requests for a workload to be scheduled, which creates one or more instances of that workload. If a client sends 3 successive requests for scheduling the same workload, the controller should store those requests and asks the scheduler to create new instances. The scheduler then updates the controller about the status of each and every instance.

The controller should not decide if a client workload is already instantiated or not, it should just serve client requests, tracks those requests and the corresponding cluster (nodes, instances) status changes.

> If not, does it mean that the controller serves only to pass workloads to scheduler, store data about states/workloads and give this data to client?

The controller is not smart. Neither is the scheduler nor the agent. Their whole and single purpose is to try and serve client requests, and, as a set of components, provide an accurate view of the cluster status.

Reply:
updated, I removed the mentions of the "scheduler comparisons"